daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -1032,46 +1116,18 @@ private[lf] final class Compiler(
           List(
             s.SCaseAlt(
               t.SCPPrimCon(PCTrue),
-              let(env, s.SEApp(s.SEBuiltin(SBCheckTemplateKey(templateId)), List(s.SEValue.Unit))) {
-                (templateKeyCheck, env) =>
-                  val contract = s.SEValue(disclosedContract.argument)
-
-                  s.SECase(
-                    env.toSEVar(templateKeyCheck),
-                    List(
-                      s.SCaseAlt(
-                        t.SCPPrimCon(PCTrue),
-                        let(
-                          env,
-                          s.SEApp(
-                            s.SEVal(ContractKeyWithMaintainersDefRef(templateId)),
-                            List(contract),
-                          ),
-                        ) { (contractPos, env) =>
-                          let(env, s.SEApp(s.SEBuiltin(SBSome), List(env.toSEVar(contractPos)))) {
-                            (optionalContractPos, env) =>
-                              s.SEApp(
-                                s.SEVal(ToCachedContractDefRef(templateId)),
-                                List(contract, env.toSEVar(optionalContractPos)),
-                              )
-                          }
-                        },
-                      ),
-                      s.SCaseAlt(
-                        t.SCPDefault,
-                        s.SEApp(
-                          s.SEVal(ToCachedContractDefRef(templateId)),
-                          List(contract, s.SEValue.None),
-                        ),
-                      ),
-                    ),
-                  )
+              checkPreCondition(
+                env,
+                templateId,
+                contract,
+              ) { (env: Env) =>
+                translateDisclosedContractWithTemplateChecked(env, templateId, disclosedContract)
               },
             ),
             s.SCaseAlt(
               t.SCPDefault,
               s.SEApp(
-                s.SEBuiltin(SBCrash(s"Template $templateId does not exist and it should")),
+                s.SEBuiltin(SBTemplateNotFound(templateId)),
```

Comment:
I thing we should avoid extending errors exposed to the client when possible.

We have already too many error, that
- make difficult to have consistent errors messages, 
- complexity documentation,
- complexity testing

In this particular case, I think the the preprocessing should check --among other things-- that the template ID exists, like that we can simply crash here with an internal error. The current implementation of `SBCrash` is good enough for that.

Reply:
No problem. I'll remove the TemplateNotFound set of changes.